numba/_lapack.c

```diff
@@ -0,0 +1,1035 @@
+ * the size of int at all the call sites.
+ */
+#define F_INT int
+
+typedef float (*sdot_t)(F_INT *n, void *dx, F_INT *incx, void *dy, F_INT *incy);
+typedef double (*ddot_t)(F_INT *n, void *dx, F_INT *incx, void *dy, F_INT
+                         *incy);
+typedef npy_complex64 (*cdot_t)(F_INT *n, void *dx, F_INT *incx, void *dy,
+                                F_INT *incy);
+typedef npy_complex128 (*zdot_t)(F_INT *n, void *dx, F_INT *incx, void *dy,
+                                 F_INT *incy);
+
+typedef void (*xxgemv_t)(char *trans, F_INT *m, F_INT *n,
+                         void *alpha, void *a, F_INT *lda,
+                         void *x, F_INT *incx, void *beta,
+                         void *y, F_INT *incy);
+
+typedef void (*xxgemm_t)(char *transa, char *transb,
+                         F_INT *m, F_INT *n, F_INT *k,
+                         void *alpha, void *a, F_INT *lda,
+                         void *b, F_INT *ldb, void *beta,
+                         void *c, F_INT *ldc);
+
+/* Vector * vector: result = dx * dy */
+NUMBA_EXPORT_FUNC(F_INT)
+numba_xxdot(char kind, char conjugate, F_INT n, void *dx, void *dy,
+            void *result)
+{
+    void *raw_func = NULL;
+    F_INT _n;
+    F_INT inc = 1;
+
+    switch (kind)
+    {
+        case 'd':
+            raw_func = get_cblas_ddot();
+            break;
+        case 's':
+            raw_func = get_cblas_sdot();
+            break;
+        case 'c':
+            raw_func = conjugate ? get_cblas_cdotc() : get_cblas_cdotu();
+            break;
+        case 'z':
+            raw_func = conjugate ? get_cblas_zdotc() : get_cblas_zdotu();
+            break;
+        default:
+        {
+            PyGILState_STATE st = PyGILState_Ensure();
+            PyErr_SetString(PyExc_ValueError,
+                            "invalid kind of *DOT function");
+            PyGILState_Release(st);
+        }
+        return -1;
+    }
+    if (raw_func == NULL)
+        return -1;
+
+    _n = (F_INT) n;
+
+    switch (kind)
+    {
+        case 'd':
+            *(double *) result = (*(ddot_t) raw_func)(&_n, dx, &inc, dy, &inc);;
+            break;
+        case 's':
+            *(float *) result = (*(sdot_t) raw_func)(&_n, dx, &inc, dy, &inc);;
+            break;
+        case 'c':
+            *(npy_complex64 *) result = (*(cdot_t) raw_func)(&_n, dx, &inc, dy,\
+                                        &inc);;
+            break;
+        case 'z':
+            *(npy_complex128 *) result = (*(zdot_t) raw_func)(&_n, dx, &inc,\
+                                         dy, &inc);;
+            break;
+    }
+
+    return 0;
+}
+
+/* Matrix * vector: y = alpha * a * x + beta * y */
+NUMBA_EXPORT_FUNC(F_INT)
+numba_xxgemv(char kind, char *trans, F_INT m, F_INT n,
+             void *alpha, void *a, F_INT lda,
+             void *x, void *beta, void *y)
+{
+    void *raw_func = NULL;
+    F_INT _m, _n;
+    F_INT _lda;
+    F_INT inc = 1;
+
+    switch (kind)
```

Comment:
We have so many occurences of this basic pattern now (from `switch(kind)` to `if (raw_func == NULL) return -1;`, included) that it may be worthwhile to create a macro for it and compress a bit of vertical space by calling the macro.